pensieve/config.py

```diff
@@ -0,0 +1,160 @@
+import enum
+from typing import Dict, List, Optional
+
+import attr
+import cattr
+import mozanalysis.metrics
+import mozanalysis.metrics.desktop
+import mozanalysis.segments
+
+import pensieve.experimenter
+
+"""
+Spec objects are direct representations of the configuration and contain unresolved references
+to metrics and data sources.
+
+Calling .resolve(config_spec) on a Spec object produces a concrete resolved Conf class.
+"""
+
+
+_converter = cattr.Converter()
+
+
+@enum.unique
+class Platform(enum.Enum):
+    DESKTOP = "desktop"
+    FENIX = "fenix"
+
+
+@attr.s(auto_attribs=True)
+class MetricReference:
+    name: str
+
+    def resolve(self, spec: "ConfigurationSpec") -> mozanalysis.metrics.Metric:
+        search = None
+        if spec.experiment.platform == Platform.DESKTOP:
+            search = mozanalysis.metrics.desktop
+
+        attr = None
+        if search and hasattr(search, self.name):
+            attr = getattr(search, self.name)
+        if self.name in spec.metrics.definitions:
+            attr = spec.metrics.definitions[self.name].resolve(spec)
+        if isinstance(attr, mozanalysis.metrics.Metric):
+            return attr
+
+        raise ValueError(f"Could not find the metric {self.name}")
+
+
+@attr.s(auto_attribs=True)
+class DataSourceReference:
+    name: str
+
+    def resolve(self, cfg: "ConfigurationSpec") -> mozanalysis.metrics.DataSource:
+        raise NotImplementedError
+
+
+@attr.s(auto_attribs=True)
+class ExperimentSpec:
+    def sync(self, experiment: pensieve.experimenter.Experiment):
+        raise NotImplementedError
+
+
+@attr.s(auto_attribs=True)
+class MetricDefinition:
```

Comment:
They're conceptually different because a definition is concrete and a reference is abstract, and I don't think we always want to let one substitute for the other -- I think it's cleaner if the `metrics.weekly` key of the config file contains only references, and not a mixture of references and inline definitions.

The properties of these classes is what defines how these entities can be represented in the configuration. All these classes do is describe how these entities are defined and resolved, so mixing entities that have to be defined and resolved differently doesn't feel right to me.

Reply:
> The properties of these classes is what defines how these entities can be represented in the configuration

This makes more sense to me after looking at how you call/use this module below.